src/imp/pidinfo.c

```diff
@@ -180,12 +276,109 @@ struct pid_info *pid_info_create (pid_t pid)
         goto err;
     if ((pi->cg_owner = path_owner (pi->cg_path)) == (uid_t) -1)
         goto err;
+    if (pid_command (pid, pi->command, sizeof (pi->command)) < 0)
+        goto err;
     return pi;
 err:
     pid_info_destroy (pi);
     return NULL;
 }
 
-/*
- * vi: ts=4 sw=4 expandtab
+int pid_kill_children_fallback (pid_t parent, int sig)
+{
+    int count = 0;
+    int rc = 0;
+    DIR *dirp = NULL;
+    struct dirent *dent;
+    pid_t pid;
+    pid_t ppid;
+
+    if (parent <= (pid_t) 0 || sig < 0) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    if (!(dirp = opendir ("/proc")))
+        return -1;
+
+    while ((dent = readdir (dirp))) {
+        if (parse_pid (dent->d_name, &pid) < 0)
+            continue;
+        if ((ppid = pid_ppid (pid)) < 0) {
+            /* ENOENT is an expected error since a process on the system
+             *  could have exited between when we read the /proc dirents
+             *  and when we are checking for /proc/PID/status.
+             */
+            if (errno != ENOENT) {
+                imp_warn ("Failed to get ppid of %d: %s",
+                          (int) pid,
+                          strerror (errno));
+                rc = -1;
+            }
+            continue;
+        }
+        if (ppid == parent && kill (pid, sig) < 0) {
+            imp_warn ("Failed to send signal %d to pid %d: %s",
+                      sig,
+                      (int) pid,
+                      strerror (errno));
+            rc = -1;
+            continue;
+        }
+        count++;
+    }
+    if (count == 0) {
+        imp_warn ("No children signaled for pid %d", (int) parent);
+        rc = -1;
+    }
+    closedir (dirp);
```

Comment:
This function appears to need to set errno when it returns -1 (callers expect it?)  but
- continued loop iteration clobbers errno in the two "continue" blocks
- errno is not set if count == 0
- closedir() clobbers errno
- imp_warn() clobbers errno